fork_choice/confirmation-rule.md

```diff
@@ -0,0 +1,574 @@
+    """ 
+    Returns the total weight of votes for this epoch from future committees after the current slot.
+    """
+    assert block_root in store.block_states
+
+    state = store.block_states[block_root]
+
+    current_slot = get_current_slot(store)
+    first_slot_next_epoch = compute_start_slot_at_epoch(Epoch(compute_epoch_at_slot(current_slot) + 1))
+    return get_committee_weight_between_slots(state, Slot(current_slot + 1), Slot(first_slot_next_epoch - 1))
+```
+
+#### `get_leaf_block_roots`
+
+```python
+def get_leaf_block_roots(store: Store, block_root: Root) -> Set[Root]:
+    children = [
+        root for root in store.blocks.keys()
+        if store.blocks[root].parent_root == block_root
+    ]
+
+    if any(children):
+        # Get leaves of all children and add to the set.
+        leaves = set().union(*[get_leaf_block_roots(store, child) for child in children])
+        return leaves
+    else:
+        # This block is a leaf.
+        return set([block_root])
+
+```
+
+#### `get_current_epoch_participating_indices`
+
+```python
+def get_epoch_participating_indices(state: BeaconState, 
+                                    active_validator_indices: Sequence[ValidatorIndex],
+                                    current_epoch: bool) -> Set[ValidatorIndex]: 
+    if current_epoch:
+        epoch_participation = state.current_epoch_participation
+    else:
+        epoch_participation = state.previous_epoch_participation
+
+    return set([
+        i for i in active_validator_indices 
+        if has_flag(epoch_participation[i], TIMELY_TARGET_FLAG_INDEX)
+    ])
+```
+
+#### `get_ffg_support`
+
+```python
+def get_ffg_support(store: Store, block_root: Root) -> Gwei:
+    """
+    Returns the total weight supporting the checkpoint in the block's chain at block's epoch.
+    """
+    block = store.blocks[block_root]
+    block_epoch = compute_epoch_at_slot(block.slot)
+    current_epoch = get_current_store_epoch(store)
+
+    # This function is only applicable to current and previous epoch blocks
+    assert current_epoch in [block_epoch, block_epoch + 1]
+
+    checkpoint_root = get_checkpoint_block(store, block_root, block_epoch)
+    checkpoint = Checkpoint(root=checkpoint_root, epoch=block_epoch)
+
+    if checkpoint not in store.checkpoint_states:
+        return Gwei(0)
+
+    checkpoint_state = store.checkpoint_states[checkpoint]
+
+    leaf_roots = get_leaf_block_roots(store, block_root)
+
+    active_validator_indices = get_active_validator_indices(checkpoint_state, block_epoch)
+    participating_indices = set().union(*[
+        get_epoch_participating_indices(
+            store.block_states[root], 
+            active_validator_indices, 
+            block_epoch == current_epoch
+        )
+        for root in leaf_roots
+    ])
+
+    return get_total_balance(checkpoint_state, participating_indices)
+```
+
+
+#### `is_ffg_confirmed`
+
+```python
+def is_ffg_confirmed(store: Store,
```

Comment:
My suggestion would be to encapsulate the support computation in a separate function and simplify the end of the `is_ffg_confirmed` function in the following way:

```python
def compute_max_honest_ffg_support(total_active_balance: Gwei,
                                   ffg_support_for_checkpoint: Gwei,
                                   remaining_ffg_weight: Gwei,
                                   confirmation_byzantine_threshold: uint64,
                                   confirmation_slashing_threshold: Gwei) -> Gwei:
    """
    Helper function computing the left side of inequality, i.e.
    S - min(alpha * W_t, S) + (1 - beta) * W_f
    """

    # min(beta * W_t, alpha * W_t, S)
    max_adversarial_ffg_support = Gwei(min(
        (total_active_balance * confirmation_byzantine_threshold - 1) // 100 + 1,
        confirmation_slashing_threshold,
        ffg_support_for_checkpoint,
    ))

    # (1 - beta) * W_f
    remaining_honest_ffg_support = Gwei(
        ((100 - confirmation_byzantine_threshold) * remaining_ffg_weight - 1) // 100 + 1)

    # S - min(alpha * W_t, S) + (1 - beta) * W_f
    return ffg_support_for_checkpoint - max_adversarial_ffg_support + remaining_honest_ffg_support


def is_ffg_confirmed(store: Store,
                     confirmation_byzantine_threshold: uint64,
                     confirmation_slashing_threshold: uint64,
                     block_root: Root) -> bool:
    ...

    max_honest_ffg_support = compute_max_honest_ffg_support(
        total_active_balance,
        ffg_support_for_checkpoint,
        remaining_ffg_weight,
        confirmation_byzantine_threshold,
        confirmation_slashing_threshold,
    )

    return 3 * max_honest_ffg_support >= 2 * total_active_balance
```

Note that the above computations uses lower precision than the original function to compute `remaining_honest_ffg_support`. But considering that this computation uses balance in Gwei the difference should be negligible (< 100 Gwei). As an upside we get a better readability which I think should be favoured here.

Also, on a related note, I think that this function should use `uint64` and its aliases like `Gwei` as considering the nature of computations (max multiplication is by a factor of 100) no overflow does seem to be possible.

Reply:
For the point about using `uint64` vs `int`, please see [this comment](https://github.com/ethereum/consensus-specs/pull/3339#discussion_r1259118938) above.